[hist] Implement AutoZoom of TH1/TH2/TH3/TAxis to filled range

ferdymercury opened the pull request:
# This Pull request:

## Changes or fixes:

Fixes https://github.com/root-project/root/issues/14538

## Checklist:

- [x] tested changes locally
- [x] updated the docs (if necessary)

ferdymercury committed:
[hist] Implement AutoZoom of TH1/TH2/TH3/TAxis to filled range

Fixes https://github.com/root-project/root/issues/14538

[hist] restore original class comment in order not to change streamerinfo checksum

[hist] prevent spurious looping on dimensions not belonging to the
derived class

as found out by linev

[hist] Split: 1 autozooms parent histogram, 1 autozooms all hists on pad

[hist] fix bug in max starting value and reset iterator before reusing

[hist] allow specifying a margin in GetRangeOfFilledWeights

[hist] adapt calls to new margin signature and add a margin of 1 bin

[hist] add safety check for static_cast

linev reviewed (changes requested):
Great!

Features was implemented long-long time ago in JSROOT,
now it is time to have it in ROOT.

But several changes are required:
- [x] exclude loop repetition in `TH1::GetRangeOfFilledWeights`
- [x] if `TH1::AutoZoom` is invoked, it only should affect that histogram and not change zooming in all other histograms in the pad
- [x] same true for `TAxis::AutoZoom` - it only should change properties of this axis and not all histograms in the pad
- [x] but must be checked with superimposed histograms, THStack, TMultiGraph - they all can have side effects
- [x] fix failures with roottest

ferdymercury commented on hist/hist/src/TH1.cxx:
```diff
@@ -7904,19 +7952,143 @@ void TH1::ResetStats()
+             }
+          }
+      }
+      return;
```
good point, will do.

ferdymercury commented on hist/hist/src/TH1.cxx:
```diff
@@ -7904,19 +7952,143 @@ void TH1::ResetStats()
+      first = start;
+      for(Int_t binx = start; binx <= lastX; binx++) {
+         for(auto biny = start; biny <= lastY; biny++) {
+            for(auto binz = start; binz <= lastZ; binz++) {
```
Good point, will fix.

ferdymercury commented:
> * if `TH1::AutoZoom` is invoked, it only should affect that histogram and not change zooming in all other histograms in the pad

This is already the behavior 'most of the time'. Only if DrawFrame was called before, the multi-analysis is run.
I was trying to mimick what Unzoom was doing, why should it be different?

linev commented:
> Only if DrawFrame was called before, the multi-analysis is run.

`DrawFrame` is special use-case and not used with `THStack` or `TMultiGraph` painting.
And `Unzoom()` can behave differently as `AutoZoom()`.

ferdymercury committed:
[test] adapt roottests to new menu functions

[test] add some extra bytes

[test] give more informative error output on stress failure

linev commented:
We need to wait with this PR until `roottest` merged into ROOT master.
Then one can sync all changes in one PR.

ferdymercury commented:
I thought it was already merged, and I had rebased and synced into one.

ferdymercury commented on roottest/root/io/filemerger/CMakeLists.txt:
```diff
@@ -335,22 +335,22 @@ if(${compression_default} STREQUAL "zlib")
             ROOTTEST_ADD_TEST(simple-lz4-compr-level9
                           PRECMD ${ROOT_hadd_CMD} -f409 hsimple409.root hsimple.root
-                          COMMAND ${ROOT_root_CMD} -q -l -b "${CMAKE_CURRENT_SOURCE_DIR}/testSimpleFile.C(\"hsimple409.root\",25000,409,517008,15)"
+                          COMMAND ${ROOT_root_CMD} -q -l -b "${CMAKE_CURRENT_SOURCE_DIR}/testSimpleFile.C(\"hsimple409.root\",25000,409,517020,15)"
```
```suggestion
                          COMMAND ${ROOT_root_CMD} -q -l -b "${CMAKE_CURRENT_SOURCE_DIR}/testSimpleFile.C(\"hsimple409.root\",25000,409,517024,15)"
```

linev replied:
Do you understand why file size in test is changing?

ferdymercury replied:
Not really (stored TStreamerInfo contains maybe the number of functions?), but it happens often with this test. See e.g. https://github.com/root-project/root/pull/15021#issuecomment-2013772175 or https://github.com/root-project/roottest/commit/46d2d618c5f32660ae452fb993e3ad0f2173718a by @pcanal

linev replied:
New methods in the class should not affect streamer infos. Small adjustment probably ok here, but I did not understand why it happens. @pcanal ?

pcanal replied:
>  but it happens often with this test.

Still it needs to be understood what is changing ....

> New methods in the class should not affect streamer infos.

Indeed, it should not.

pcanal replied:
> Still it needs to be understood what is changing ....

The last one I investigated turned out that it was cause by the inadvertent loading (i.e. bug) of many unwanted libraries.  I.e. it could be serious ... or it might not :)

ferdymercury replied:
Resolved: we found out that changing comment of a persistent class member is "dangerous" since it changes the StreamerInfo.

pcanal replied:
It is okay to improve the class member documentation.  This is a valid reason to have to update the expected file size.

ferdymercury replied:
(I did that on Sergei's request). Question: does changing the streamerinfo imply that the ClassDef version must be increased?

linev replied:
I think it is better not to change streamer infos (including comments) without real need.
And class version is not need to be changed if only comments are changing.

linev commented:
> I thought it was already merged

Yes, now I also see `roottest` in main repository. Things changing very fast!

ferdymercury commented on hist/hist/src/TAxis.cxx:
```diff
@@ -1328,6 +1328,104 @@ void TAxis::UnZoom()
+      }
+   }
+
+   hobj1->GetRangeOfFilledWeights(dim, first, last, false);
```
```suggestion
   hobj1->GetRangeOfFilledWeights(dim, first, last, 1, false);
```

ferdymercury commented on hist/hist/src/TAxis.cxx:
```diff
@@ -1328,6 +1328,104 @@ void TAxis::UnZoom()
+      TH1 *hobj = static_cast<TH1*>(obj);
+      if (dim > hobj->GetDimension())
+         continue;
+      hobj->GetRangeOfFilledWeights(dim, first, last, false);
```
```suggestion
      hobj->GetRangeOfFilledWeights(dim, first, last, 1, false);
```

ferdymercury committed:
[hist] support THStack Autozoom

ferdymercury commented:
> 1. For 1-D histogram with few entries  `TH1::AutoZoom()` changes zooming for Y axis which is not nice.

I guess that's a feature of "TH1::UnZoom" that is also not nice, that could be handled in a separate PR ?

> 2\. Selected range is exact position of the non-zero bins. If peak is very sharp and has just single bin width - produced result is unusable. One need some default margins on the left and right side of the peak.

I have implemented this now. It could be modified to accept a relative percentage rather than a fixed number of bins if wanted.

> 3\. If there are several peaks - AutoZoom always show all peaks and not most significant one.

I do not understand what you mean. I am proposing AutoZoom to show all peaks, rather than the most significant.

> 3\. Also if histogram already pre-zoomed to some range - peak should be searched in this range.

I kind of disagree, sometimes I like to zoom in to see a specific thing, and then AutoZoom to zoom out to the filled range. Maybe ZoomToFit is a better word for the function??

> With superimposed histograms `AutoZoom` works only when first histogram is selected - for all others histogram method has no effect.

I had made a mistake. Is now fixed, thanks for finding out. You need though to click on the x-axis and click on AutoZoomAll, not AutoZoom. An equivalent function can be created in the TH1::AutoZoomAll if you think it's worth.

> With `THStack` method does not work.

Support has been added now, via TAxis::AutozoomAll.

ferdymercury commented on test/stress.cxx:
```diff
@@ -366,7 +366,7 @@ void stress2()
 #else
 #ifdef R__HAS_CLOUDFLARE_ZLIB
-      Long64_t lastgood = 9813;
+      Long64_t lastgood = 9826; // changes for TH1 PR#18459
```
```suggestion
      Long64_t lastgood = 9813;
```

ferdymercury commented on roottest/root/io/filemerger/CMakeLists.txt:
```diff
@@ -75,7 +75,7 @@ if(ZLIB_CF)
   ROOTTEST_ADD_TEST(merged-zlib
                     PRECMD  ${ROOT_hadd_CMD} -f101 mzlibfile1-4.root mzlibfile1-2.root mzlibfile3-4.root
-                    COMMAND ${ROOT_root_CMD} -q -l -b "${CMAKE_CURRENT_SOURCE_DIR}/testMergedFile.C(\"mzlibfile1-4.root\",101,5014,20)"
+                    COMMAND ${ROOT_root_CMD} -q -l -b "${CMAKE_CURRENT_SOURCE_DIR}/testMergedFile.C(\"mzlibfile1-4.root\",101,5020,20)"
```
```suggestion
                    COMMAND ${ROOT_root_CMD} -q -l -b "${CMAKE_CURRENT_SOURCE_DIR}/testMergedFile.C(\"mzlibfile1-4.root\",101,5014,20)"
```

ferdymercury commented on roottest/root/io/filemerger/CMakeLists.txt:
```diff
@@ -222,17 +222,17 @@ if(${compression_default} STREQUAL "zlib")
       ROOTTEST_ADD_TEST(simple-lz4-compr-level9
                      PRECMD ${ROOT_hadd_CMD} -f409 hsimple409.root hsimple.root
-                     COMMAND ${ROOT_root_CMD} -q -l -b "${CMAKE_CURRENT_SOURCE_DIR}/testSimpleFile.C(\"hsimple409.root\",25000,409,517008,15)"
+                     COMMAND ${ROOT_root_CMD} -q -l -b "${CMAKE_CURRENT_SOURCE_DIR}/testSimpleFile.C(\"hsimple409.root\",25000,409,517020,15)"
```
```suggestion
                     COMMAND ${ROOT_root_CMD} -q -l -b "${CMAKE_CURRENT_SOURCE_DIR}/testSimpleFile.C(\"hsimple409.root\",25000,409,517008,15)"
```

ferdymercury commented on roottest/root/io/filemerger/CMakeLists.txt:
```diff
@@ -222,17 +222,17 @@ if(${compression_default} STREQUAL "zlib")
       ROOTTEST_ADD_TEST(simple-lz4-compr-level1
                      PRECMD ${ROOT_hadd_CMD} -f401 hsimple401.root hsimple.root
-                     COMMAND ${ROOT_root_CMD} -q -l -b "${CMAKE_CURRENT_SOURCE_DIR}/testSimpleFile.C(\"hsimple401.root\",25000,401,519976,5)"
+                     COMMAND ${ROOT_root_CMD} -q -l -b "${CMAKE_CURRENT_SOURCE_DIR}/testSimpleFile.C(\"hsimple401.root\",25000,401,520007,5)"
```
```suggestion
                     COMMAND ${ROOT_root_CMD} -q -l -b "${CMAKE_CURRENT_SOURCE_DIR}/testSimpleFile.C(\"hsimple401.root\",25000,401,519976,5)"
```

ferdymercury commented on roottest/root/io/filemerger/CMakeLists.txt:
```diff
@@ -325,7 +325,7 @@ if(${compression_default} STREQUAL "zlib")
          ROOTTEST_ADD_TEST(simplek-lz4-compr-level4
                      PRECMD ${ROOT_hadd_CMD} -fk404 hsimpleK404.root hsimpleK209.root hsimpleK.root hsimple.root hsimple9.root hsimple209.root hsimple409.root hsimple92.root
-                     COMMAND ${ROOT_root_CMD} -q -l -b "${CMAKE_CURRENT_SOURCE_DIR}/testSimpleFile.C(\"hsimpleK404.root\",24*25000,404,12047492,30)"
+                     COMMAND ${ROOT_root_CMD} -q -l -b "${CMAKE_CURRENT_SOURCE_DIR}/testSimpleFile.C(\"hsimpleK404.root\",24*25000,404,12047512,30)"
```
```suggestion
                     COMMAND ${ROOT_root_CMD} -q -l -b "${CMAKE_CURRENT_SOURCE_DIR}/testSimpleFile.C(\"hsimpleK404.root\",24*25000,404,12047492,30)"
```

ferdymercury commented on roottest/root/io/filemerger/CMakeLists.txt:
```diff
@@ -335,22 +335,22 @@ if(${compression_default} STREQUAL "zlib")
          ROOTTEST_ADD_TEST(simple-lz4-compr-level4
                         PRECMD ${ROOT_hadd_CMD} -f404 hsimple404.root hsimple.root
-                        COMMAND ${ROOT_root_CMD} -q -l -b "${CMAKE_CURRENT_SOURCE_DIR}/testSimpleFile.C(\"hsimple404.root\",25000,404,517262,15)"
+                        COMMAND ${ROOT_root_CMD} -q -l -b "${CMAKE_CURRENT_SOURCE_DIR}/testSimpleFile.C(\"hsimple404.root\",25000,404,517292,15)"
```
```suggestion
                        COMMAND ${ROOT_root_CMD} -q -l -b "${CMAKE_CURRENT_SOURCE_DIR}/testSimpleFile.C(\"hsimple404.root\",25000,404,517262,15)"
```

ferdymercury commented on roottest/root/io/filemerger/CMakeLists.txt:
```diff
@@ -335,22 +335,22 @@ if(${compression_default} STREQUAL "zlib")
             ROOTTEST_ADD_TEST(simple-lz4-compr-level9
                           PRECMD ${ROOT_hadd_CMD} -f409 hsimple409.root hsimple.root
-                          COMMAND ${ROOT_root_CMD} -q -l -b "${CMAKE_CURRENT_SOURCE_DIR}/testSimpleFile.C(\"hsimple409.root\",25000,409,516975,30)"
+                          COMMAND ${ROOT_root_CMD} -q -l -b "${CMAKE_CURRENT_SOURCE_DIR}/testSimpleFile.C(\"hsimple409.root\",25000,409,517004,30)"
```
```suggestion
                          COMMAND ${ROOT_root_CMD} -q -l -b "${CMAKE_CURRENT_SOURCE_DIR}/testSimpleFile.C(\"hsimple409.root\",25000,409,516975,30)"
```

ferdymercury commented on roottest/root/io/filemerger/CMakeLists.txt:
```diff
@@ -335,22 +335,22 @@ if(${compression_default} STREQUAL "zlib")
             ROOTTEST_ADD_TEST(simple-lz4-compr-level9
                           PRECMD ${ROOT_hadd_CMD} -f409 hsimple409.root hsimple.root
-                          COMMAND ${ROOT_root_CMD} -q -l -b "${CMAKE_CURRENT_SOURCE_DIR}/testSimpleFile.C(\"hsimple409.root\",25000,409,517008,15)"
+                          COMMAND ${ROOT_root_CMD} -q -l -b "${CMAKE_CURRENT_SOURCE_DIR}/testSimpleFile.C(\"hsimple409.root\",25000,409,517024,15)"
```
```suggestion
                          COMMAND ${ROOT_root_CMD} -q -l -b "${CMAKE_CURRENT_SOURCE_DIR}/testSimpleFile.C(\"hsimple409.root\",25000,409,517008,15)"
```

ferdymercury commented on roottest/root/io/filemerger/CMakeLists.txt:
```diff
@@ -335,22 +335,22 @@ if(${compression_default} STREQUAL "zlib")
          ROOTTEST_ADD_TEST(simple-lz4-compr-level1
                         PRECMD ${ROOT_hadd_CMD} -f401 hsimple401.root hsimple.root
-                        COMMAND ${ROOT_root_CMD} -q -l -b "${CMAKE_CURRENT_SOURCE_DIR}/testSimpleFile.C(\"hsimple401.root\",25000,401,519976,10)"
+                        COMMAND ${ROOT_root_CMD} -q -l -b "${CMAKE_CURRENT_SOURCE_DIR}/testSimpleFile.C(\"hsimple401.root\",25000,401,520007,10)"
```
```suggestion
                        COMMAND ${ROOT_root_CMD} -q -l -b "${CMAKE_CURRENT_SOURCE_DIR}/testSimpleFile.C(\"hsimple401.root\",25000,401,519976,10)"
```

ferdymercury commented on roottest/root/io/filemerger/CMakeLists.txt:
```diff
@@ -368,5 +368,5 @@ endif()
 ROOTTEST_ADD_TEST(simple-lzma-compr-level9
                   PRECMD ${ROOT_hadd_CMD} -f209 hsimple209.root hsimple.root
-                  COMMAND ${ROOT_root_CMD} -q -l -b "${CMAKE_CURRENT_SOURCE_DIR}/testSimpleFile.C(\"hsimple209.root\",25000,209,392709,17)"
+                  COMMAND ${ROOT_root_CMD} -q -l -b "${CMAKE_CURRENT_SOURCE_DIR}/testSimpleFile.C(\"hsimple209.root\",25000,209,392716,17)"
```
```suggestion
                  COMMAND ${ROOT_root_CMD} -q -l -b "${CMAKE_CURRENT_SOURCE_DIR}/testSimpleFile.C(\"hsimple209.root\",25000,209,392709,17)"
```

linev commented:
Still, AutoZoom on simple TH1 with few entries changes Y-zoom,
And current range is not taken into account when search for the range.

ferdymercury commented:
> Still, AutoZoom on simple TH1 with few entries changes Y-zoom

But Unzoom does the same, it's changing that too. It's zooming in rather than out.

> And current range is not taken into account when search for the range.

Yes, but I argue that that's intended. Autozoom in matplotlib works also like that.

linev commented:
> But Unzoom does the same, it's changing that too. It's zooming in rather than out.

Then one need to fix `Unzoom` too. It is not nice behavior.

> Autozoom in matplotlib works also like that.

Use-case is simple. One makes approx zooming with mouse and then call method to get peak fully displayed.
Therefore behavior of matplotlib is not best approach here.

linev commented on hist/hist/src/TAxis.cxx:
```diff
@@ -1270,7 +1270,7 @@ void TAxis::UnZoom()
    //unzoom object owning this axis
    SetRange(0,0);
-   TH1 *hobj1 = (TH1*)GetParent();
+   TH1 *hobj1 = GetParent()->InheritsFrom(TH1::Class()) ? static_cast<TH1*>(GetParent()) : nullptr;
```
why not simply `dynamic_cast<TH1*>(GetParent())`?
Or you also should check that GetParent() is not `nullptr`

ferdymercury commented on hist/hist/src/TAxis.cxx:
```diff
@@ -1269,7 +1270,7 @@ void TAxis::UnZoom()
    //unzoom object owning this axis
    SetRange(0,0);
-   TH1 *hobj1 = (TH1*)GetParent();
+   TH1 *hobj1 = GetParent()->InheritsFrom(TH1::Class()) ? static_cast<TH1*>(GetParent()) : nullptr;
```
```suggestion
   TH1 *hobj1 = GetParent() && GetParent()->InheritsFrom(TH1::Class()) ? static_cast<TH1*>(GetParent()) : nullptr;
```

ferdymercury commented:
> Then one need to fix `Unzoom` too. It is not nice behavior.

Last commit tries to fix that.

dpiparo commented:
@couet @linev is this ready to be merged?

linev commented:
@dpiparo No, there several issues with the PR.

ferdymercury commented:
In my opinion this is ready to be merged.

The remaining comments by @linev are debatable from my point of view. For example, I disagree with the feature that zooms to a visible peak even if there might be higher peaks outside the currently visible range, since it goes against the typical behavior of eg matplotlib. If this is really wanted, then this should be a separate function that could be implemented in a separate PR. It could be called Smartzoom insted of Autozoom.
